crates/full-node/sov-sequencer/src/preferred/replica/event_receiver.rs

```
@@ -1,178 +1,109 @@
+    InvalidEventSequence(Option<EventType>, EventType),
 }
 
 pub(crate) struct EventReceiver {
     connection_string: String,
     db_data_sender: tokio::sync::mpsc::Sender<DbData>,
-    db_data_receiver: tokio::sync::mpsc::Receiver<DbData>,
     shutdown_sender: watch::Sender<()>,
+    listener: PgListener,
+    query_pool: PgPool,
 }
 
 impl EventReceiver {
-    pub(crate) async fn new(connection_string: String, shutdown_sender: watch::Sender<()>) -> Self {
+    pub(crate) async fn new(
+        connection_string: String,
+        shutdown_sender: watch::Sender<()>,
+    ) -> (Self, tokio::sync::mpsc::Receiver<DbData>) {
         let (db_data_sender, db_data_receiver) = tokio::sync::mpsc::channel(PAGE_SIZE as usize);
-        Self {
-            connection_string,
-            db_data_sender,
-            db_data_receiver,
-            shutdown_sender,
-        }
-    }
 
-    pub(crate) async fn spawn_db_data_fetcher(&mut self) -> JoinHandle<()> {
         // Create a separate persistent connection pool for querying transaction data
         let query_pool = match PgPoolOptions::default()
             .max_connections(5) // Small pool since we're just doing simple queries
-            .connect(&self.connection_string)
+            .connect(&connection_string)
             .await
         {
             Ok(pool) => pool,
             Err(e) => {
                 error!("Failed to connect to PostgreSQL: {e:?}. Replica shutting down.");
-                exit_rollup(&self.shutdown_sender).await;
-                unreachable!("EventReceiver: impossible happened rollup didn't exit");
-            }
-        };
-
-        let mut start_event_id = match latest_event_id(&query_pool).await {
-            Ok(latest_event_id) => latest_event_id,
-            Err(e) => {
-                error!("Failed to get latest event id: {e:?}. Replica shutting down.");
-                exit_rollup(&self.shutdown_sender).await;
+                exit_rollup(&shutdown_sender).await;
```

As per discussion on standup, can we return an error here from the task and have the process exit correctly? Non-blocking

---

I would do it in separate PR for all the relevant places.